corehq/apps/locations/forms.py

```diff
@@ -248,3 +255,62 @@ def save(self, instance=None, commit=True):
             pass
 
         return location
+
+
+class UsersAtLocationForm(MultipleSelectionForm):
+    def __init__(self, domain_object, location, *args, **kwargs):
+        self.domain_object = domain_object
+        self.location = location
+        super(UsersAtLocationForm, self).__init__(
+            initial={'selected_ids': self.users_at_location},
+            *args, **kwargs
+        )
+
+    @property
+    @memoized
+    def users_at_location(self):
+        user_query = (UserES()
+                      .domain(self.domain_object.name)
+                      .mobile_users()
+                      .location(self.location._id)
+                      .fields([]))
+        return user_query.run().doc_ids
+
+    def already_have_locations(self, users):
+        user_query = (UserES()
+                      .domain(self.domain_object.name)
+                      .mobile_users()
+                      .doc_id(list(users))
+                      .exists('location_id')
+                      .fields(['username']))
+        return [raw_username(u['username']) for u in user_query.run().hits]
+
+    def unassign_users(self, users):
+        for doc in iter_docs(CommCareUser.get_db(), users):
+            # This could probably be sped up by bulk saving, but there's a lot
+            # of stuff going on - seems tricky.
+            CommCareUser.wrap(doc).unset_location()
```

Comment:
another approach here would be to run it asynchronously via celery so the slowness isn't visible to the user (similar to how we do user deletions). not sure whether that will impact the UX though.